support-lib/cpp/Future.hpp

```diff
@@ -0,0 +1,307 @@
+    Future(Future&& other) noexcept = default;
+    Future& operator= (Future&& other) noexcept = default;
+    // not copyable
+    Future(const Future& other) = delete;
+    Future& operator= (const Future& other) noexcept = delete;
+    // Future becomes invalid after `then()` is called on it
+    bool isValid() const {
+        return std::atomic_load(&_sharedState) != nullptr;
+    }
+    // returns true if the result can be `get()` without blocking
+    bool isReady() const {
+        auto sharedState = std::atomic_load(&_sharedState);
+        assert(sharedState);    // call on invalid future will trigger assertion
+        std::unique_lock lk(sharedState->mutex);
+        return sharedState->isReady();
+    }
+    // wait until future becomes `isReady()`
+    void wait() const {
+        auto sharedState = std::atomic_load(&_sharedState);
+        assert(sharedState);    // call on invalid future will trigger assertion
+        std::unique_lock lk(sharedState->mutex);
+        sharedState->cv.wait(lk, [state = sharedState] {return state->isReady();});
+    }
+    // wait until future becomes `isReady()` and return the result. This can
+    // only be called once.
+    auto get() {
+        detail::SharedStatePtr<T> sharedState;
+        sharedState = std::atomic_exchange(&_sharedState, sharedState);
+        assert(sharedState);    // call on invalid future will trigger assertion
+        std::unique_lock lk(sharedState->mutex);
+        sharedState->cv.wait(lk, [state = sharedState] {return state->isReady();});
+        if (!sharedState->exception) {
+            return sharedState->getValueUnsafe();
+        } else {
+            std::rethrow_exception(sharedState->exception);
+        }
+    }
+    // If at the moment of calling `then()`, the result (or exception) is
+    // already available, then the handler routine will immediately be called in
+    // the current thread. Returns a new future that wraps the return value of
+    // the handler routine. The current future becomes invalid after this call.
+    template<typename FUNC>
+    auto then(FUNC&& handler) {
+        using HandlerReturnType = std::invoke_result_t<FUNC, Future<T>>;
+        detail::SharedStatePtr<T> sharedState;
+        sharedState = std::atomic_exchange(&_sharedState, sharedState);
+        assert(sharedState);    // a second call will trigger assertion
+        auto nextPromise = std::make_unique<Promise<HandlerReturnType>>();
+        auto nextFuture = nextPromise->getFuture();
+        auto continuation = [handler = std::forward<FUNC>(handler), nextPromise = std::move(nextPromise)] (detail::SharedStatePtr<T> x) mutable {
+            try {
+                if constexpr(std::is_void_v<HandlerReturnType>) {
+                    handler(Future<T>(x));
+                    nextPromise->setValue();
+                } else {
+                    nextPromise->setValue(handler(Future<T>(x)));
+                }
+            } catch (std::exception& e) {
```

Comment:
Can we make this a const reference? If only to signal that nothing will modify `e`.